### DISCUSSION TITLE
meaning of pg.compute_bootci() with custom function

### DISCUSSION THREAD

**Discussion by FlorinAndrei** [raphaelvallat/pingouin#274] [Q&A] [answered]
This is the dataset:

[BirdPecks.csv](https://github.com/raphaelvallat/pingouin/files/8915350/BirdPecks.csv)

Prep work:

```
BirdPecks = pd.read_csv("BirdPecks.csv")

Nboot = 10000

pg1 = BirdPecks[BirdPecks["group"] == 1]["pecks"].to_numpy()
pg2 = BirdPecks[BirdPecks["group"] == 2]["pecks"].to_numpy()
pg1_shape = pg1.shape[0]
pg2_shape = pg2.shape[0]

sns.boxplot(data=BirdPecks, x="group", y="pecks");
```

![pecks_hist](https://user-images.githubusercontent.com/901867/173983759-44f7e629-229e-494a-ab1a-679c5240a645.png)

I am trying to bootstrap a confidence interval for the difference in means between pg1 and pg2. Here's the "manual" version:

```
bd = []
for i in tqdm(range(Nboot)):
    res_pg1 = np.random.choice(pg1, size=pg1_shape, replace=True, p=None)
    res_pg2 = np.random.choice(pg2, size=pg2_shape, replace=True, p=None)
    bd.append(np.mean(res_pg1) - np.mean(res_pg2))

sns.histplot(bd)
```

![man_hist](https://user-images.githubusercontent.com/901867/173984054-d878828c-09c8-475b-9228-9a5c6c6b7343.png)

Here's what I think is the equivalent via Pingouin:

```
ci, dist = pg.compute_bootci(
    x=pg1,
    y=pg2,
    func=lambda x, y: np.mean(x) - np.mean(y),
    confidence=0.9,
    n_boot=Nboot,
    paired=False,
    return_dist=True
)
print(ci)

sns.histplot(dist);
```

![pg_hist](https://user-images.githubusercontent.com/901867/173984221-23a09ab5-1e12-4648-8f24-f8f69084ebe2.png)

Why are the histograms so different?

The distribution returned by `pg.compute_bootci()` - is it not the distribution of the output of the `func` parameter?

**Comment by FlorinAndrei** [comment]
Another question is that the confidence interval returned by `pg.compute_bootci()` varies a lot if I use different options for the `method` parameter. I'm not sure why there is so much change. And they all differ quite a lot from:

```
pg.ttest(pg1, pg2, confidence=0.9)
```

Which gives:

```
               T        dof alternative     p-val  \
T-test  1.570018  29.902445   two-sided  0.126933   

                                          CI90%   cohen-d   BF10     power  
T-test  [-2.618045665165672, 67.13569272398922]  0.549579  0.851  0.323706 
```

In the past, I've noticed `method="cper"` gives the closest results to `boot.ci()` from R (used with `type="bca"`). But here the results from PG do not match at all what I get from R.

**Comment by raphaelvallat** [comment]
@FlorinAndrei please see the source code of Pingouin here:

https://github.com/raphaelvallat/pingouin/blob/8825a215f50a21fdaa53aa63fc25a1af44aee7bc/pingouin/effsize.py#L343-L363

A few differences:
1. Unlike your code, Pingouin replaces the pairs (Xi, Yi) and not Xi and Yi separately
2. Because of this, `n` will be the size of the smallest array: https://github.com/raphaelvallat/pingouin/blob/8825a215f50a21fdaa53aa63fc25a1af44aee7bc/pingouin/effsize.py#L298

As for the different confidence intervals, please refer to the source code, which should hopefully be self-explanatory:

https://github.com/raphaelvallat/pingouin/blob/8825a215f50a21fdaa53aa63fc25a1af44aee7bc/pingouin/effsize.py#L365-L393

For the unit tests of the confidence intervals against Matlab, see:

https://github.com/raphaelvallat/pingouin/blob/8825a215f50a21fdaa53aa63fc25a1af44aee7bc/pingouin/tests/test_effsize.py#L80-L94

**Comment by FlorinAndrei** [comment]
Quick question: if `paired=False` then why replacements are still done paired? My understanding for the generic bootstrap is that you resample each group separately - at least if no pairing is involved (with pairing I am not sure how bootstrap is supposed to work, perhaps it's supposed to be how PG is doing it?).

**Comment by raphaelvallat** [comment, marked as answer]
@FlorinAndrei you are right — I have just looked at the implementation of the recently-added [scipy.stats.bootstrap](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html) and they indeed do resample each group separately when `paired=False`, which makes more sense because otherwise you're discarding data if `x` and `y` do not have the same length. My implementation was based on Matlab's [bootci](https://www.mathworks.com/help/stats/bootci.html) which only support the case where `x` and `y` have the same length: 

> When you use multiple data input arguments d1,...,dN, you can specify some arguments as scalar values, but all nonscalar arguments must have the same number of rows.

I'll open an issue for this. Thanks
